### PR TITLE
EOS-16883: I added single SW upgrade ISO bundle structure validation

### PR DESCRIPTION
**Pull request by dmitrydb**
I added base classes for single SW upgrade ISO bundle structure validation

### CODE REVIEW

File: api/python/provisioner/commands/validator/validator.py [hunk 1]
```diff
@@ -0,0 +1,250 @@
...
+from abc import abstractmethod
+from pathlib import Path
+from typing import Dict, Callable
+
+from ...errors import ValidationError
+
+logger = logging.getLogger(__name__)
+
+
+class SchemeValidator:
```
Comment (andkononykhin2) [inline api/python/provisioner/commands/validator/validator.py:11]: didn't you forget to inherit form `ABC` here?

File: api/python/provisioner/commands/validator/validator.py [hunk 2]
```diff
@@ -0,0 +1,250 @@
...
+
+        Raises
+        ------
+        ValidationError
+            If validation is failed.
+        """
+        pass
+
+
+class Validator:
```
Comment (andkononykhin2) [inline api/python/provisioner/commands/validator/validator.py:42]: How it differs from the `SchemeValidator` class?

**Review by ypise — APPROVED** [review]

File: api/python/provisioner/commands/validator/validator.py [hunk 3]
```diff
@@ -0,0 +1,250 @@
...
+        permissions: int/str
+            requested file permissions
+        owner: str
+            requested file owner
+        group: str
+            requested file group
+        is_symlink: bool
+            defines if file should be a symlink
+        """
+        self._required = required
```
Comment (andkononykhin2) [inline api/python/provisioner/commands/validator/validator.py:98]: I would suggest to use `attr` helpers here like in other API parts:
- please use `attr.ib` (if actually needed) to declare attrs
- for all validator you may actually use attr validators so validation would during both validator object creation time and during `validator` method call
- many of validations you do here are already available: attr [built-in validators](https://www.attrs.org/en/stable/api.html#validators) and provisioner custom [ones](https://github.com/Seagate/cortx-prvsnr/blob/bf833997fa2bd383e9109f5e748538c04acab327/api/python/provisioner/utils.py#L44-L72))
- for more complicated validators you can write your own as a separate function in the module
- also you can consider to use [converters](https://www.attrs.org/en/stable/init.html#converters) and make you input assumptions more flexible (e.g. `Union[str, Path]`), like [here](https://github.com/Seagate/cortx-prvsnr/blob/bf833997fa2bd383e9109f5e748538c04acab327/api/python/provisioner/paths.py#L27)
- docstring can be attached to class level summary, like [here](https://github.com/rackerlabs/otter/blob/ec0a76d25dcbb50636dfc9316674715e138b9d0c/otter/convergence/selfheal.py#L27-L58)

To summarize:
- for object level attributes (`__init__` args) I suggest to use `attr` API
- for `validator` method you can re-use the same validator functions (isinstance, Path exists ...)
- converters would help you as well

Finally the module (currently about 250 lines) might be condensed into 100 lines or even less.

**Review by andkononykhin2 — CHANGES_REQUESTED** [review]
@dmitrydb The approach looks promising. Thanks. Please check my comments regarding the implementation.

File: api/python/provisioner/commands/validator/validator.py (hunk 3 above)
Comment (dmitrydb) [reply api/python/provisioner/commands/validator/validator.py]: I used the `attr` module only for validators attributes declaration. `validate`-method is only one method implemented for class-validators. It implements the basic and simple logic for scheme validation. And it can be used for many catalog structure validation at the same time. Also, it is the best way to implement nested files/catalog traversal.
So let's leave `attr` for class-instance and model initialization, data validation and conversion routines. Also, I base `RELEASE.INFO` and other content type validation mostly on the `attr` module. Because it is the good choice for that. For scheme validation (catalog/file structure validation) provided logic is simple and flexible. Inside of `validate` method I have access to all validator attributes and can use it to build complex validations. The base logic takes around 50 lines of code, other lines are docstrings and `attr` annotations.

File: api/python/provisioner/commands/validator/validator.py (hunk 2 above)
Comment (dmitrydb) [reply api/python/provisioner/commands/validator/validator.py]: I planned to distinguish these two classes. `SchemeValidator` is a complex class-validator which used only for different scheme validations. For this moment I left only `Validator` base-class.

File: api/python/provisioner/commands/validator/validator.py (hunk 1 above)
Comment (dmitrydb) [reply api/python/provisioner/commands/validator/validator.py]: Yes, it should be inherit from `ABC`. I fixed that.

File: api/python/provisioner/commands/set_swupgrade_repo.py [hunk 4]
```diff
@@ -167,7 +207,21 @@ def dynamic_validation(self, params: inputs.SWUpgradeRepo, targets: str):  # noq
...
+                                                    SW_UPGRADE_BUNDLE_SCHEME)
+
+            try:
+                catalog_scheme_validator.validate(iso_mount_dir)
+            except ValidationError as e:
+                logger.debug("Catalog structure validation error occurred: "
+                             f"{e}")
+                raise SWUpdateRepoSourceError(
+                                        str(repo.source),
+                                        "Catalog structure validation error "
```
[the remark this answers is not in the document]
Comment (dmitrydb) [reply api/python/provisioner/commands/set_swupgrade_repo.py]: I fixed that where is possible. In some places error messages too long. I think we should not spend a log of time improving string concatenation.

File: api/python/provisioner/utils.py [hunk 5]
```diff
@@ -60,6 +60,25 @@ def converter_path(value):
     return value if value is None else Path(str(value))
 
 
+def file_scheme_key_converter(value: dict):
```
Comment (andkononykhin2) [inline api/python/provisioner/utils.py:63]: could you rename to `converter_file_scheme_key` to be aligned with existent code?

**Review by andkononykhin2 — CHANGES_REQUESTED** [review]
@dmitrydb Good improvement. I've noticed few less critical things. Please address them as well.

File: api/python/provisioner/commands/set_swupgrade_repo.py [hunk 6]
```diff
@@ -25,14 +25,54 @@
...
+from .validator import FileValidator, DirValidator, FileSchemeValidator
 
 logger = logging.getLogger(__name__)
 
 
+# NOTE: keys can be either Path object or strings
+SW_UPGRADE_BUNDLE_SCHEME = {
+    '3rd_party': DirValidator(
+        {
+            "THIRD_PARTY_RELEASE.INFO": FileValidator(required=True),
```
[the remark this answers is not in the document]
Comment (dmitrydb) [reply api/python/provisioner/commands/set_swupgrade_repo.py]: This moment `RELEASE.INFO` file for `3rd_party` repo names `THIRD_PARTY_RELEASE.INFO`. I have to create the new constant for that. I am not sure that this file will be renamed in future by RE guys.

File: api/python/provisioner/commands/set_swupgrade_repo.py [hunk 7]
```diff
@@ -25,14 +25,54 @@
...
+            "repodata": DirValidator(
```
[the remark this answers is not in the document]
Comment (dmitrydb) [reply api/python/provisioner/commands/set_swupgrade_repo.py]: I make class-alias for that. Let's consider to create the separate ticket to add to `YumRepoDataValidator` additional functionality.

File: api/python/provisioner/commands/set_swupgrade_repo.py [hunk 8]
```diff
@@ -167,13 +207,27 @@ def dynamic_validation(self, params: inputs.SWUpgradeRepo, targets: str):  # noq
 
             iso_mount_dir = base_dir / REPO_CANDIDATE_NAME
 
-            release_file = f'{iso_mount_dir}/{RELEASE_INFO_FILE}'
+            catalog_scheme_validator = FileSchemeValidator(
```
[the remark this answers is not in the document]
Comment (dmitrydb) [reply api/python/provisioner/commands/set_swupgrade_repo.py]: Fixed that. But honestly say it is a matter of dispute.

File: api/python/provisioner/commands/set_swupgrade_repo.py [hunk 9]
```diff
@@ -167,13 +207,27 @@ def dynamic_validation(self, params: inputs.SWUpgradeRepo, targets: str):  # noq
...
+
+            release_file = (f'{iso_mount_dir}/{CORTX_ISO_DIR}/'
+                            f'{RELEASE_INFO_FILE}')
             try:
                 metadata = load_yaml(release_file)
             except Exception as exc:
                 raise SWUpdateRepoSourceError(
-                    str(repo.source),
-                    f"Failed to load '{RELEASE_INFO_FILE}' file: {exc}"
+                            str(repo.source),
```
[the remark this answers is not in the document]
Comment (dmitrydb) [reply api/python/provisioner/commands/set_swupgrade_repo.py:224]: Additional indentation is needed for better visual perception. Again, it is a matter of dispute. This code is already in PEP8 formatted code. For me current variant is more readable.

File: api/python/provisioner/commands/validator/validator.py [hunk 10]
```diff
@@ -0,0 +1,252 @@
...
+            requested file group
+        is_symlink: bool
+            defines if file should be a symlink
+
+    At initialization time `attr.s` is responsible for `_required`
+    and `_content_validator` attributes validation.
+    """
+
+    _required: bool = attr.ib(
+        validator=attr.validators.optional(attr.validators.instance_of(bool)),
```
[the remark this answers is not in the document]
Comment (dmitrydb) [reply api/python/provisioner/commands/validator/validator.py]: I agree, fixed that

File: api/python/provisioner/commands/validator/validator.py [hunk 11]
```diff
@@ -0,0 +1,252 @@
...
+        default=False
+    )
+    _content_validator: Callable[[Path], None] = attr.ib(
```
[the remark this answers is not in the document]
Comment (dmitrydb) [reply api/python/provisioner/commands/validator/validator.py]: Ok, I improved that

File: api/python/provisioner/commands/validator/validator.py [hunk 12]
```diff
@@ -0,0 +1,252 @@
...
+            if self._required:
+                raise ValidationError(reason=f"File '{path}' should exist.")
+            return
+
+        logger.debug(f"File '{path}' exists.")
+
+        if not path.is_file():
+            raise ValidationError(reason=f"'{path}' is not a regular file.")
+
+        if self._content_validator and callable(self._content_validator):
```
[the remark this answers is not in the document]
Comment (dmitrydb) [reply api/python/provisioner/commands/validator/validator.py]: Yes, I forgot to delete this previous time. It is fixed

File: api/python/provisioner/commands/validator/validator.py [hunk 13]
```diff
@@ -0,0 +1,252 @@
...
+            # Should raise an Exception if validation fails
+            self._content_validator(path)
```
[the remark this answers is not in the document]
Comment (dmitrydb) [reply api/python/provisioner/commands/validator/validator.py]: Thank you for workaround. I used it to make codacy "happy"

File: api/python/provisioner/commands/validator/validator.py [hunk 14]
```diff
@@ -0,0 +1,252 @@
...
+    At initialization time `attr.s` is responsible for `_file_scheme`
+    validation and converting keys of file scheme from string to `Path`
+    instance.
+    """
+
+    _files_scheme: Dict = attr.ib(
+        validator=attr.validators.optional(
+            attr.validators.deep_mapping(
+                key_validator=attr.validators.instance_of((str, Path)),
+                # we can't refer to DirValidator here.
```
[the remark this answers is not in the document]
Comment (dmitrydb) [reply api/python/provisioner/commands/validator/validator.py]: Originally, it was supposed to be like
```python
value_validator=attr.validators.instance_of((FileValidator, DirValidator)),
```
Since I started developing the new Validators for validation, it is better to leave only parent class here.

File: api/python/provisioner/commands/validator/validator.py [hunk 15]
```diff
@@ -0,0 +1,252 @@
...
+            if self._required:
+                raise ValidationError(reason=f"File '{path}' should exist.")
+            return
+
+        logger.debug(f"Directory '{path}' exists.")
+
+        if not path.is_dir():
+            raise ValidationError(reason=f"'{path}' is not a directory")
+
+        if self._files_scheme and isinstance(self._files_scheme, dict):
```
[the remark this answers is not in the document]
Comment (dmitrydb) [reply api/python/provisioner/commands/validator/validator.py]: I agree, fixed that

File: api/python/provisioner/commands/validator/validator.py [hunk 16]
```diff
@@ -0,0 +1,252 @@
...
+    -----
+    At initialization time `attr.s` is responsible for `_scheme` validation
+    and converting keys of file scheme from string to `Path` instance.
+    """
+
+    _scheme: Dict = attr.ib(
+        validator=attr.validators.optional(
+            attr.validators.deep_mapping(
+                key_validator=attr.validators.instance_of((str, Path)),
+                value_validator=attr.validators.instance_of((FileValidator,
```
[the remark this answers is not in the document]
Comment (dmitrydb) [reply api/python/provisioner/commands/validator/validator.py:221]: Since I started developing the new Validators for validation, it is better to leave only parent class here. I changed that

File: api/python/provisioner/utils.py (hunk 5 above)
Comment (dmitrydb) [reply api/python/provisioner/utils.py]: Yes, sure. Fixed that.

File: api/python/provisioner/commands/set_swupgrade_repo.py (hunk 6 above)
[the remark this answers is not in the document]
Comment (dmitrydb) [reply api/python/provisioner/commands/set_swupgrade_repo.py]: Yes, we have two option here - add the special validator here, or use the `content_validator` attribute to set validation function. We can assume that `ReleaseInfoValidator = FileValidator(required=True, content_validator=release_info_validator)`

File: api/python/provisioner/commands/set_swupgrade_repo.py (hunk 7 above)
[the remark this answers is not in the document]
Comment (andkononykhin2) [reply api/python/provisioner/commands/set_swupgrade_repo.py]: sure, could you take care about that? Thanks

File: api/python/provisioner/commands/set_swupgrade_repo.py (hunk 6 above)
[the remark this answers is not in the document]
Comment (andkononykhin2) [reply api/python/provisioner/commands/set_swupgrade_repo.py]: Yes. We can proceed with any option for now. Just some thoughts:
- a separate class will help to keep things organized better
- we might need to support multiple versions of release info structure, so class might be used to hide that logic inside

If you are not going to cover that in nearest PR please crate some task for that. Thanks

File: api/python/provisioner/commands/set_swupgrade_repo.py (hunk 8 above)
[the remark this answers is not in the document]
Comment (andkononykhin2) [reply api/python/provisioner/commands/set_swupgrade_repo.py]: Sure.

File: api/python/provisioner/commands/validator/validator.py [hunk 17]
```diff
@@ -0,0 +1,257 @@
...
+
+    At initialization time `attr.s` is responsible for `_required`
+    and `_content_validator` attributes validation.
+    """
+
+    _required: bool = attr.ib(
+        validator=attr.validators.instance_of(bool),
+        default=False
+    )
+    _content_validator: Optional[Callable[[Path], None]] = attr.ib(
```
Comment (andkononykhin2) [inline api/python/provisioner/commands/validator/validator.py:77]: `Optional` itself implies `None` so should be: `Optional[Callable[Path]]`

File: api/python/provisioner/commands/set_swupgrade_repo.py (hunk 9 above)
[the remark this answers is not in the document]
Comment (andkononykhin2) [reply api/python/provisioner/commands/set_swupgrade_repo.py:224]: but it is a legacy code and how many such additional spaces are enough?

File: api/python/provisioner/utils.py [hunk 18]
```diff
@@ -60,6 +60,25 @@ def converter_path(value):
...
+    ----------
+    value: dict
+        a given file scheme which represents some catalog structure
+
+    Returns
+    -------
+    dict
+        new dictionary object where keys are `Path` instances
+    """
+    return value if value is None else {converter_path(k): v for k, v in
```
Comment (andkononykhin2) [inline api/python/provisioner/utils.py:78]: makes sense to consider only defined (non None) values here and deal with None outside (e.g. `attr.validators.optional`) but it's an improvement task since would require review of the many places in the legacy code

File: api/python/provisioner/commands/set_swupgrade_repo.py [hunk 19]
```diff
@@ -191,10 +240,13 @@ def dynamic_validation(self, params: inputs.SWUpgradeRepo, targets: str):  # noq
...
                 except KeyError:
                     raise SWUpdateRepoSourceError(
-                        str(repo.source),
-                        f"No release data found in '{RELEASE_INFO_FILE}'"
+                            str(repo.source),
+                            f"No release data found in '{RELEASE_INFO_FILE}'"
                     )
 
+            else:
+                logger.info("Catalog structure validation succeeded")
```
Comment (andkononykhin2) [inline api/python/provisioner/commands/set_swupgrade_repo.py:248]: that thing should be seemed as a subpart of previous logger.info message which is about the whole stage of repo validation, if you have some quick ideas how to make it clearer please apply.
I see options:
- add `\t` at the beginning
- add some context of the bundle under validation

**Review by andkononykhin2 — APPROVED** [review]
@dmitrydb Thank you. Good contribution.
